std/experimental/perpetual/perpetual.d

```diff
@@ -0,0 +1,352 @@
+/**
+ * Map variable, POD structure or array to file
+ *   to make it shared and persistent between program invocations.
+ * Macros:
+  * Source:    $(PHOBOSSRC std/perpetual.d)
+ */
+module std.perpetual;
+//import std.stdio;
+
+private import std.file;
+private import core.stdc.stdio;
+private import std.traits;
+private import std.string;
+private import std.conv;
+private import std.algorithm;
+private import std.exception;
+version (Windows) {
+private import core.sys.windows.windows;
+private import std.utf;
+private import std.windows.syserror;
+} else version (Posix) {
+private import core.sys.posix.fcntl;
+private import core.sys.posix.unistd;
+private import core.sys.posix.sys.mman;
+private import core.sys.posix.sys.stat;
+} else {
+	static assert(0);
+}
+
+
+
+/**
+ * Persistently maps value type object to file
+ */
+class Perpetual(T)
+{
+	version (Windows) {
+
+	} else version (Posix) {
+		private int _fd;
+		private void* _map;
+	} else
+		static assert(0);
+
+	enum _tag="Perpetual!("~T.stringof~")";
+
+	static if(is(T == Element[],Element)) {
+	// dynamic array 
+		private size_t _size=0;
+		enum bool dynamic=true;
+		static assert(!hasIndirections!Element, Element.stringof~" is reference type");
+		@property T Ref() const { return cast(T) _map[0.._size]; }
+
+	} else {
+	// value type
+		enum bool dynamic=false;
+		static assert(!hasIndirections!T, T.stringof~" is reference type");
+		@property ref T Ref() const {	return *cast(T*)(_map); }
+	}
+
+/**
+ * Get reference to wrapped object.
+ */
+	alias Ref this;
+
+	private bool _owner=true;
+	@property auto master() { return _owner; }
+
+/**
+ * Return string representation for wrapped object instead of self.
+ */
+	override string toString() const { return to!string(Ref); }
+
+
+
+/**
+ * Open file and assosiate object with it.
+ * The file is extended if smaller than requred. Initialized
+ *   with T.init if created or extended.
+ */
+	this(string path) {
+		map(path);
+		static if(dynamic) {
+			// we don't need initialization here, the file already exists
+			//   so, is to be initialized
+		} else if(master) {
+			*cast(T*)(_map)=T.init;
+		}
+ 	}
+
+
+
+/**
+ * Unmap the file and release resources.
+ * The file is left on the disk and may be reopened.
+ */
+	~this() {
+		unmap();
+	}
+
+
+
+/**
+ * Flash memory to disc.
+ */
+	void sync() {
```

Comment:
May want to expose some of version(Windows)/version(Posix) functions as stand-alone portable functions that pass around some opaque struct that describes the mapping. Then std.mmfile would use them too and the Q about reusing std.mmfile would be addressed.